nat/service_ics.go

```diff
@@ -0,0 +1,224 @@
+func (nat *serviceICS) Disable() (resErr error) {
+	for iface, rule := range nat.ifaces {
+		if err := nat.Del(rule); err != nil {
+			log.Errorf("%s Failed to cleanup internet connection sharing for '%s' interface: %v", natLogPrefix, iface, err)
+			if resErr != nil {
+				resErr = err
+			}
+		}
+	}
+
+	_, err := powerShell("Set-Service -Name RemoteAccess -StartupType " + nat.remoteAccessStatus)
+	if err != nil {
+		err = errors.Wrap(err, "failed to revert RemoteAccess service startup type")
+		log.Errorf("%s %v", natLogPrefix, err)
+		if resErr != nil {
+			resErr = err
+		}
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		err = errors.Wrap(err, "failed to get public interface name")
+		log.Errorf("%s %v", natLogPrefix, err)
+		if resErr != nil {
+			resErr = err
+		}
+	}
+
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+	if err != nil {
+		err = errors.Wrap(err, "failed to disable internet connection sharing")
+		log.Errorf("%s %v", natLogPrefix, err)
+		if resErr != nil {
+			resErr = err
+		}
+	}
+
+	return resErr
+}
+
+func getInterfaceBySubnet(subnet string) (string, error) {
+	_, ipnet, err := net.ParseCIDR(subnet)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse subnet from request")
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get a list of network interfaces")
+	}
+
+	for _, iface := range ifaces {
+		addrs, err := iface.Addrs()
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get list of interface addresses")
+		}
+
+		for _, addr := range addrs {
+			var ip net.IP
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+
+			if ipnet.Contains(ip) {
+				return iface.Name, nil
+			}
+		}
+	}
+	return "", errors.New("interface not found")
+}
+
+func getPublicInterfaceName() (string, error) {
+	out, err := powerShell(`Get-WmiObject -Class Win32_IP4RouteTable | where { $_.destination -eq '0.0.0.0' -and $_.mask -eq '0.0.0.0'} | foreach { $_.InterfaceIndex }`)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get interface from the default route")
+	}
+
+	ifaceID, err := strconv.Atoi(strings.TrimSpace(string(out)))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse interface ID")
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get a list of network interfaces")
+	}
+
+	for _, iface := range ifaces {
+		if iface.Index == ifaceID {
+			return iface.Name, nil
+		}
+	}
+
+	return "", errors.New("interface not found")
+}
+
+func powerShell(cmd string) ([]byte, error) {
```

Comment:
Is it possible to mock `exec.Command` and have some test with common situations simulated?

Reply:
Yeah, we can do some mocked tests.

Reply:
Added some testing for it.